src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -326,108 +312,73 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+                if (lengthProperty is not null)
+                {
+                    lengthAccess = new BoundPropertyAccess(node, receiverPlaceholder, lengthProperty, LookupResultKind.Viable, lengthProperty.Type) { WasCompilerGenerated = true };
+                }
+                else
+                {
+                    lengthAccess = new BoundBadExpression(node, LookupResultKind.Empty, ImmutableArray<Symbol?>.Empty, ImmutableArray<BoundExpression>.Empty, CreateErrorType(), hasErrors: true) { WasCompilerGenerated = true };
                 }
-                lookupResult.Clear();
             }
+            else
+            {
+                hasErrors |= !TryBindLengthOrCount(node, receiverPlaceholder, out lengthAccess, bindingDiagnostics);
+            }
+
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var systemIndexType = GetWellKnownType(WellKnownType.System_Index, bindingDiagnostics, node);
+            argumentPlaceholder = new BoundListPatternIndexPlaceholder(node, systemIndexType) { WasCompilerGenerated = true };
+            analyzedArguments.Arguments.Add(argumentPlaceholder);
+
+            indexerAccess = BindElementAccessCore(node, receiverPlaceholder, analyzedArguments, bindingDiagnostics).MakeCompilerGenerated();
+            indexerAccess = CheckValue(indexerAccess, BindValueKind.RValue, bindingDiagnostics);
+            Debug.Assert(indexerAccess is BoundIndexerAccess or BoundIndexOrRangePatternIndexerAccess or BoundArrayAccess or BoundBadExpression or BoundDynamicIndexerAccess);
+            analyzedArguments.Free();
 
-            // If the argType is missing or the indexer lookup has failed, we will fallback to the implicit indexer support.
-            found = TryLookupLengthOrCount(receiverType, lookupResult, out lengthProperty, ref useSiteInfo) &&
-                    TryFindIndexOrRangeIndexerPattern(lookupResult, receiverOpt: null, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo);
+            diagnostics.AddRangeAndFree(bindingDiagnostics);
 
-done:
-            if (found)
+            if (!systemIndexType.HasUseSiteError)
             {
-                Debug.Assert(indexerAccess is not null ^ patternSymbol is not null);
-                Debug.Assert(lengthProperty is not null);
-                // At this point we have succeeded to bind a viable indexer,
-                // report additional binding diagnostics that we have seen so far
-                diagnostics.AddRange(bindingDiagnostics);
-                diagnostics.Add(syntax, useSiteInfo);
+                // Check required well-known member. They may not be needed
+                // during lowering, but it's simpler to always require them to prevent
+                // the user from getting surprising errors when optimizations fail
+                _ = GetWellKnownTypeMember(WellKnownMember.System_Index__op_Implicit_FromInt32, diagnostics, syntax: node);
```

Comment:
>_ = GetWellKnownTypeMember(WellKnownMember.System_Index__op_Implicit_FromInt32, diagnostics, syntax: node); [](http://example.com/codeflow?start=16&length=107)

This doesn't look consistent with lowering. Perhaps we should bind the conversion and capture the BoundConversion node. Should be fine to follow up in a separate PR.

Reply:
This is how the rest of the range binding/lowering works (it checks for some members that may or may not get used). I can file an issue if you'd like, but don't intend to spend the time on it.

Reply:
> but don't intend to spend the time on it.

That is fine.

> This is how the rest of the range binding/lowering works (it checks for some members that may or may not get used).

Just to clarify. What bothers me here is the fact that lowering never uses this member explicitly, it relies on conversion classification instead, etc. So, we are making an assumption that all that complicated machinery ends up using the method. However, there is no actual guarantee that that would be the case. For example, if method is not marked with special name flag, it won't be considered as a valid conversion operator. 

> I can file an issue if you'd like

It would be great if you could do that and assign it to me.

Reply:
Filed [https://github.com/dotnet/roslyn/issues/57825](https://github.com/dotnet/roslyn/issues/57825)